Implement support for S3 based storage providers

igorski opened the pull request:
See https://github.com/igorski/bitmappery/issues/25

* Implement AWS SDK to connect to S3 storage (should also work for S3 compatible alternatives)
* Allow configuration of third party service providers using .env files
* Allow lazy third party service authentication directly from save modal window

igorski commented on .env:
```diff
@@ -0,0 +1,53 @@
+# ---------------------------------
```
Template configuration file.

All third party storage integrations can now be enabled/disabled by specifying/omitting their API keys.

igorski commented on README.md:
```diff
@@ -116,7 +116,11 @@ create a custom handler inside `@/src/factories/action-factory` to avoid code du
+storage provider. Currently, there is support for [Dropbox](https://www.dropbox.com/developers/apps), [Google Drive](https://console.cloud.google.com/) and S3 Buckets.
+
+You can enable each of these integrations by providing the required key values for your configuration
+by creating a local `env.local`-file which will contain your custom files. You can create this
```
Copy `.env` as `env.local` and store the local configurations in the local file (it is under .gitignore as we shouldn't expose these keys).

Ange-Cesari commented:
Hi there,

I tried to look as hard as I could before my answer.

I created a `docker-compose` to be able to launch a minio as simply as possible.

Pre-requesites : 

- git
- unix/linux system arround you can use for bash
- docker & docker-compose installed and usable on your machine

All you have to do is to git clone the bitmappery project, go to the aws branch, create a new docker image. (the steps are written already in the readme.md of the project)

Once the docker image is tagged with "bitmappery-aws" (or you can edit in the docker compose)

you create a file "docker-compose.yml" and you drag the rest in: 

```yaml
#Written by Ange-Cesari
version: "3"
services:

  bitmappery:
    image: bitmappery-aws
    ports:
      - "5173:5173"
    env_file: ./env.local
    networks:
      - stack


  minio:
    image: docker.io/minio/minio
    ports:
      - "9000:9000"
      - "9001:9001"
    entrypoint: sh
    command: -c 'minio server --console-address ":9001" /data/minio'
    deploy:
      restart_policy:
        condition: on-failure
    volumes:
      - storage-data:/data
    environment:
      MINIO_ROOT_USER: 'minio'
      MINIO_ROOT_PASSWORD: 'miniostorage'
    healthcheck:
      test: ["CMD", "curl", "-f", "http://localhost:9000/minio/health/live"]
      interval: 30s
      timeout: 20s
      retries: 3
    networks:
      - stack



volumes:
  storage-data:

networks:
  stack:
    
```

Just make a docker-compose up, the dependencies will be pulled and then there will be a minio on ports 9000 and 9001, and the bitmappery on port 5173. Be aware that the port `5173` is unused on the machine, otherwise the port 5174 will have to be exposed instead, in the dockerfile and in the docker-compose.yml.

Normally the bitmappery page is accessible via localhost:5173 (as with a classic docker run) but now minio is accessible via `localhost:9001`. Connect to the minio : The credentials are in the environment variables of the docker-compose.yml, namely: 


id: `minio`
pw: `miniostorage`

Once connected, you create a bucket and you call it "bitmappery-bucket"

Once done, you go into access key, and you create an access key. Be careful to go to the end of the creation step to recover the access key and the secret key.

Once everything is validated, we go back to the terminal where we launched the docker compose and we do a:

`docker-compose down`

once everything is down, 

copy the .env file to env.local and fill in the environment variables
Once finished, we up with a `docker compose up`.

For minio, all the steps are absolutely identical to what I did with outline. Check for detailed step by step guide : https://medium.com/@ange.cesari/how-to-deploy-outline-wiki-fully-self-hosted-no-tls-with-docker-compose-e50b57baaf73

You can now try to go to `localhost:5173`

FYI, the env.local looks like that to me : 

```
VITE_S3_ACCESS_KEY=accesskey
VITE_S3_SECRET_KEY=secret access key
VITE_S3_BUCKET_NAME=bitmappery-bucket
VITE_S3_REGION=local
VITE_S3_BUCKET_URL=http://localhost:9000/bitmappery-bucket
```

you'll notice that whenever you are doing a request to the S3 bucket, you'll have a subdomain This is due to the fact that it's how a S3 bucket on AWS works. On a local host, it can be a source of problem. you could try change your /etc/hosts and add bitmappery-bucket.localhost as just localhost but I don't think it's viable and this direction shouldn't be taken. 



Normally, there should be an error on creating and recovering file from the bucket. I noticed there was an extra "/" in the URL after the bucket name. This may possibly explain the error but I think there is something else, open the browser debugger to see the queries.


In addition, to be sure that this is not a problem of minio, I recovered and modified a script to avoid having to install the client minio (command line).

Just create a file on a unix (you can use linux subsystem on windows if you want) , fill in the necessary information, make it executable. Normally we should be able to send a file to the bucket. :


```bash
#!/bin/bash

# usage: ./minio-upload my-bucket my-file.zip

bucket=$1
file=$2

host=localhost:9000
s3_key='your access key'
s3_secret='your secret key'

resource="/${bucket}/${file}"
content_type="application/octet-stream"
date=`date -R`
_signature="PUT\n\n${content_type}\n${date}\n${resource}"
signature=`echo -en ${_signature} | openssl sha1 -hmac ${s3_secret} -binary | base64`

curl -v -X PUT -T "${file}" \
          -H "Host: $host" \
          -H "Date: ${date}" \
          -H "Content-Type: ${content_type}" \
          -H "Authorization: AWS ${s3_key}:${signature}" \
          http://$host${resource} -k
``` 

just create a random file, chmod +x the script and `./script.sh bitmappery-bucket myfile`

You should have a similar answer : 

```
*   Trying 127.0.0.1:9000...
* Connected to localhost (127.0.0.1) port 9000 (#0)
> PUT /bitmappery-bucket/test.txt HTTP/1.1
> Host: localhost:9000
> User-Agent: curl/7.81.0
> Accept: */*
> Date: Sun, 23 Apr 2023 17:45:30 +0200
> Content-Type: application/octet-stream
(your keys) behind
> Authorization: AWS :
> Content-Length: 17
> Expect: 100-continue
>
* Mark bundle as not supporting multiuse
< HTTP/1.1 100 Continue
* We are completely uploaded and fine
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
< Accept-Ranges: bytes
< Content-Length: 0
< Content-Security-Policy: block-all-mixed-content
< ETag: "42992704d61965ec286zgegd63550ce8a0fe"
< Server: MinIO
< Strict-Transport-Security: max-age=31536000; includeSubDomains
< Vary: Origin
< Vary: Accept-Encoding
< X-Amz-Id-2: e3b0c44298fc1c149afbf4cfze8996fb92427ae41e4649b934ca495991b7852b855
< X-Amz-Request-Id: 17589ADBF819D598
< X-Content-Type-Options: nosniff
< X-Xss-Protection: 1; mode=block
< Date: Sun, 23 Apr 2023 15:45:30 GMT
<
```

If you have any clue, please don't hesitate to share them, i'll try them out.

igorski commented:
Hi @Ange-Cesari thanks for the testing and the Docker compose files.

I am currently away from any configuration powerful enough to host Docker for another week, but I did find a [helpful example](https://northflank.com/guides/connect-nodejs-to-minio-with-tls-using-aws-s3) that uses the AWS SDK to connect to MinIO. I noticed that the client registration step was different to the pure-AWS one.

I have updated the service connector to follow the same mechanism. Perhaps this work for your you too ?

Ange-Cesari commented:
Hey @igorski,

Thank you for the update. Unfortunately, it didn't fix the problem but the error changed.

Today, the error seemed to be the region that wasn't initialized, even though it was correctly entered and parsed from the `env.local` file.

Just so you know what I did 

I pulled the changes from your commit, I created a new docker image with the exact same name as the one before, I forced the no cache to build a fully new image so there is no issue with any layer built.

I followed the exact same step. Just so you know, doing a docker-compose down do not erase any information when they are stored, which is the case for minio. It's stored inside the container. I do have the exact same configuration as I did before doing my git pull to retrieve the changes. 

As I previously did the check-ups about the minio server (with my script that could send files) I can be sure that everything is still fine as it was for the last test.


The test I performed : 

- I tried to create within the UI a folder in my bucket S3 
- I tried to load (import) files from my bucket S3
- I tried to save files within existing folder I have access to inside the minio, to be sure there is no creation conflict.

Everytime, I get the error "Region is missing" even if it's correctly parsed from the `env.local`.

As it's not an AWS S3, region is not "us-east" etc.

The region is usually defined by `local` when using minio but I have no idea what your framework is using as "local" region. It may vary, but for outline wiki, it was "local" as region.

Here are the 2 screens I could get : 

![1](https://user-images.githubusercontent.com/46873683/235255968-db298be0-df44-4398-8125-2755d57f78d9.jpg)
![2](https://user-images.githubusercontent.com/46873683/235255960-0fa43e39-cfd1-4005-ad1e-20e63d24b8ea.jpg)

If you want to fix the code, I'll be glad to try and give you feedbacks.

Hope this helps,

Ange

igorski commented:
Hi @Ange-Cesari ,

Excuse the delay as other things got in the way.
You are right, the region is necessary, defining it as "local" did the trick.

Can you verify whether pulling the latest changes also works for you?

Ange-Cesari commented:
Hi @igorski 

I pulled your changes
When running without docker compose, directly from the code with 

Building with : 
```bash
yarn run build
```

Starting with :

```bash
yarn run dev
```

or

```bash
yarn run dev --host
```

In both cases, the "import from s3" button doesn't even show up, and nowhere in the menu there is something related to S3. the `env.local` is common to both the docker compose AND the files because the docker compose is executed where the files are built with yarn in local, it doesn't have anything to do with contenerization. What is strange, is in the browser console there is absolutely no error or warning.

In docker compose now : 

I rebuilt the docker image (still `bitmappery-aws`), nothing in my docker compose changed.


I Started the docker compose with the new built image.
runnin `localhost:5173` as always, the page shows up and I have the possibility to `import from s3` with the orange button.

I click (still watching the browser console)
Some errors but I thinkg the problem is that the key is mixed ip with the bucket name, because in the error format I got that `there is no such key` with the key being `bitmappery-bucket`

which is the name of my bucket in minio.

The "bucket-name" is still appearing as "bitmappery-bucket" so the resource became : 

```
<Error>
<Code>NoSuchKey</Code>
<Message>The specified key does not exist.</Message>
<Key>bitmappery-bucket/</Key>
<BucketName>bitmappery-bucket</BucketName>
<Resource>/bitmappery-bucket/bitmappery-bucket/</Resource>
<RequestId>17633F390DE73161</RequestId>
<HostId>REDACTED</HostId>
</Error>
```

I don't know how to help you farther than that with this because there is no really "hints". My minio is working fine, the docker compose is correct. What is more disturbing is the fact that it doesn't work without docker for the S3, using the same env.local, using the same code.

I exec in the container and verified the VITE_S3 variables and yes, everything is in the container.


Here is the error : 
![bitmappery-notworking](https://github.com/igorski/bitmappery/assets/46873683/89466900-6a11-4161-8ed1-ef08d78a3660)


Hope it helps you.

igorski commented:
Hi @Ange-Cesari 

> In both cases, the "import from s3" button doesn't even show up, and nowhere in the menu there is something related to S3. the env.local is common to...

Strange, but are you sure `env.local` isn't a typo ? It must be `.env.local` (note the first dot), I believe I made a typo somewhere in the first commits of this pull request as I stumbled upon the same problem while debugging a while back.

I have updated the pull request, I managed to have success using a local MinIO instance and both AWS S3 providers (despite this type of storage being a standard, there is definitely a difference in the way paths are handled). Could you verify whether you will have success using the latest changes ?

Ange-Cesari commented:
Hi @igorski 

Solved the problem of connecting to the bucket. 

I think there was both code that was different from the classic AWS but also a typo problem with the .env.local

However, at the interface level there is a display problem. The thumbnails do not load, and the path inside the bucket does not load, so that we cannot see all the files inside the S3 bucket (while the keys give full access to the bucket). We stay stuck at the root of the bucket

![image](https://github.com/igorski/bitmappery/assets/46873683/76210edd-aca5-4179-ae5b-7a845d6919c5)


![image](https://github.com/igorski/bitmappery/assets/46873683/5fbef154-8895-484f-9aff-a8e5ebcd185f)


![image](https://github.com/igorski/bitmappery/assets/46873683/7f064eec-524e-4d17-ae47-1046d032f8c8)


Maybe if it's not already done you should consider add some folder icon etc.

Moreover there is no output or error of any kind in the dev tools so I can't help you further with this matter but if you want me to test more feature regarding minio (for example a fix of the UI) please feel free I still do have the project running.

it's a good patch work you did there, well done!

Best, 
Ange

igorski commented:
Hi @Ange-Cesari 

Thanks for testing and the feedback. I noticed discrepancies between the way AWS and MinIO handle the paths, the solution is a bit hackish, but basically internally I align the path formatting, I verified this to work for both AWS and MinIO buckets.

As for the lack of thumbnails, yes that was interesting as that was deliberate but I omitted to communicate this 😅 
The reason is that on S3 you don't get preview thumbnails "for free". But I gave it some thought, you can add the line:

```
VITE_S3_GENERATE_THUMBNAILS=true
```

to your `.env.local` and thumbnails will now be generated on the fly. The behaviour is that BitMappery will check if a thumbnail file exists for an image. If not, it will download the image in full, generate the thumbnail and save it in S3. The next time, the operation will be a lot cheaper as the thumbnail will be served. For a local/on premise hosted MinIO solution this shouldn't be an issue from a billing perspective and only first viewing of folders will be slower (showing loaders while the thumbnails are generated).

Ange-Cesari commented:
Hey @igorski 

>Thanks for testing and the feedback. I noticed discrepancies between the way AWS and MinIO handle the paths, the solution is a bit hackish, but basically internally I align the path formatting, I verified this to work for both AWS and MinIO buckets.

Youi're welcome.

>As for the lack of thumbnails, yes that was interesting as that was deliberate but I omitted to communicate this 😅
The reason is that on S3 you don't get preview thumbnails "for free". But I gave it some thought, you can add the line:

VITE_S3_GENERATE_THUMBNAILS=true
to your .env.local and thumbnails will now be generated on the fly. The behaviour is that BitMappery will check if a thumbnail file exists for an image. If not, it will download the image in full, generate the thumbnail and save it in S3*. The next time, the operation will be a lot cheaper as the thumbnail will be served. For a local/on premise hosted MinIO solution this shouldn't be an issue from a billing perspective and only first viewing of folders will be slower (showing loaders while the thumbnails are generated).

*when deleting a file from the BitMappery UI, the associated thumbnail will also be deleted

Unfortunately it didn't change anything,
![image](https://github.com/igorski/bitmappery/assets/46873683/5649925f-b232-443c-8aa3-9bd004419843)


I don't have anything in the console and nothing wrong either in the network apart some 304 errors.
![image](https://github.com/igorski/bitmappery/assets/46873683/ac276f01-6525-4077-b555-708133c7e4f1)



![image](https://github.com/igorski/bitmappery/assets/46873683/26098f00-c568-48ec-be99-7ae8b16204e7)


![image](https://github.com/igorski/bitmappery/assets/46873683/919a8571-d121-492f-a720-bf041ab17187)


(I rebuilt a docker image to take your changes, so no problem about that, + i added your line in my .env.local)

And I tried to send in minio some files and it works, from my shell script and from your UI, and I can retrieve them without any problem. Also, your app works just fine as the other day, which means that nothing else is broken but the thumbnail feature isn't working properly somewhere but does not generate any error.

If you have any idea don't hesitate,
Best,
Ange

igorski commented:
Well, I am stumped 🙈 

I use the same Docker image for MinIO as you supplied in your examples, so I think we can rule out versioning issues... also with S3 SDK...

It looks like in your screenshots, the keys are not recognized as either files or folders (the folder icon would've worked along, even prior to my last changes). Question is why... 

If you create a BitMappery document and save it in a newly created folder in your storage root (for instance with name 'foo' to be stored in folder `/new`, which you can both specify in the save modal window), is the directory recognized correctly (e.g. shows icon) and does upon navigation, the file appear showing just `foo.bpy`* as its name ?

_*(note there's currently no preview for .bpy documents, which will be  a future improvement in general)._

Ange-Cesari commented:
Well as you can see here : 

![image](https://github.com/igorski/bitmappery/assets/46873683/ff464b64-9ff0-417f-bf4a-4b1f11e52f43)

It doesn't change anything. I don't think it comes from the container because i copied everything from your project. So as the entire folder is replicated, unless you did an absolute path somewhere I don't think it is container related.

Also, nothing related here but I happen to have click everywhere on an image i imported from s3 bucket and it appears that I might have do something with the layers, and when I opened the console to debug the icon, it just ran into that : 

![image](https://github.com/igorski/bitmappery/assets/46873683/b5312168-f8f3-41b9-b35c-52440a224233)


IMO it's not really important I just wanted to let you know it exists. However, on the icons parts, maybe doing some debug might lead to something new but I have no idea what to debug because nothing is in the console. Maybe you can try adding a lot of debug related to the keys, only for this branch, so I can try to test everything with the minio and see if the keys are loaded correctly or not..

if you have better suggestion, reach me out.

Best,
Ange

igorski commented:
Hi @Ange-Cesari,

Maybe a good way to debug is if I could see what the format is of the returned data.
If you open the network panel, filter by XHR and click the "Import from S3"-button, the first GET request should show a list of files in your storage root.

You can obfuscate the folder and file names, but I would like to see all trailing and leading slashes.
Basically the problem comes from separating the file key into a folder and file name structure, so it looks like for some reason you are getting a list that is structured differently.

For instance, a snapshot like below would suffice (it lists file keys that are inside a "sub directory"):

![list_bucket](https://github.com/igorski/bitmappery/assets/730069/1b7fdcb7-79e4-4051-80c9-e396f5ec9779)

Ange-Cesari commented:
Hey @igorski 

I don't wanted to hide anything for the debug so I deleted the previous personal files and I put new one without personal information. As the instance of minio is local and not exposed on Internet, I can show everything.

Here is a screen of what I got : 

![image](https://github.com/igorski/bitmappery/assets/46873683/792719e3-a7a1-43d0-9f68-defc297e2aa7)


```xml

<?xml version="1.0" encoding="UTF-8"?>
<ListBucketResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/">
    <Name>bitmappery-bucket</Name>
    <Prefix></Prefix>
    <KeyCount>6</KeyCount>
    <MaxKeys>500</MaxKeys>
    <Delimiter></Delimiter>
    <IsTruncated>false</IsTruncated>
    <Contents>
        <Key>Koala_in_weightless_position_wearing_a_blue_overall_si_c7d8acd6-e35c-4559-b8b8-d066a46fedf0.bpy</Key>
        <LastModified>2023-07-02T07:33:53.285Z</LastModified>
        <ETag>&#34;a343a825346f55ea0aaa46400f61726c-1&#34;</ETag>
        <Size>1832856</Size>
        <Owner>
            <ID>02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4</ID>
            <DisplayName>minio</DisplayName>
        </Owner>
        <StorageClass>STANDARD</StorageClass>
    </Contents>
    <Contents>
        <Key>foo/Koala_in_weightless_position_wearing_a_blue_overall_si_c7d8acd6-e35c-4559-b8b8-d066a46fedf0-removebg-preview.bpy</Key>
        <LastModified>2023-07-02T07:34:43.775Z</LastModified>
        <ETag>&#34;b5ab72a6040546a0063eba4bf1d1252f-1&#34;</ETag>
        <Size>438932</Size>
        <Owner>
            <ID>02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4</ID>
            <DisplayName>minio</DisplayName>
        </Owner>
        <StorageClass>STANDARD</StorageClass>
    </Contents>
    <Contents>
        <Key>foo/Koala_in_weightless_position_wearing_a_blue_overall_si_c7d8acd6-e35c-4559-b8b8-d066a46fedf0.bpy</Key>
        <LastModified>2023-07-02T07:34:04.885Z</LastModified>
        <ETag>&#34;a343a825346f55ea0aaa46400f61726c-1&#34;</ETag>
        <Size>1832856</Size>
        <Owner>
            <ID>02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4</ID>
            <DisplayName>minio</DisplayName>
        </Owner>
        <StorageClass>STANDARD</StorageClass>
    </Contents>
    <Contents>
        <Key>foo/abstract_logo_of_a_koalas_head_black_and_whit_with_enc_1d6aaa36-633b-43bf-ae75-13b8390fd954.bpy</Key>
        <LastModified>2023-07-02T07:34:33.495Z</LastModified>
        <ETag>&#34;121a02f75bad6046f2de87df35ac76d2-1&#34;</ETag>
        <Size>2067756</Size>
        <Owner>
            <ID>02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4</ID>
            <DisplayName>minio</DisplayName>
        </Owner>
        <StorageClass>STANDARD</StorageClass>
    </Contents>
    <Contents>
        <Key>test.txt</Key>
        <LastModified>2023-04-23T16:26:11.312Z</LastModified>
        <ETag>&#34;42992704d61965ec286d63550ce8a0fe&#34;</ETag>
        <Size>17</Size>
        <Owner>
            <ID>02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4</ID>
            <DisplayName>minio</DisplayName>
        </Owner>
        <StorageClass>STANDARD</StorageClass>
    </Contents>
    <Contents>
        <Key>test/</Key>
        <LastModified>2023-06-21T20:30:41.449Z</LastModified>
        <ETag>&#34;d41d8cd98f00b204e9800998ecf8427e&#34;</ETag>
        <Size>0</Size>
        <Owner>
            <ID>02d6176db174dc93cb1b899f7c6078f08654445fe8cf1b6ce98d8855f66bdbf4</ID>
            <DisplayName>minio</DisplayName>
        </Owner>
        <StorageClass>STANDARD</StorageClass>
    </Contents>
</ListBucketResult>

```

Tell me if you need anything else, like putting some images in some folder or something.

Best,
Ange

igorski commented:
Hi @Ange-Cesari ,

Thanks for the example output. Strange that the path styles are different than to my MinIO instance... Anyways, I have adjusted the code to not rely on upfront determination of what path styles to use, but to interpret it at runtime (it also cleaned up the code nicely).

I have verified it to work for AWS S3, my local MinIO and for your use case I have used unit tests to verify whether the files are interpreted correctly.

Could you give it another spin after pulling the latest changes ? 👼

Ange-Cesari commented:
Hey @igorski ,

You did a great job for the integration of the clickable paths.

The folder icon also works, unfortunately the preview still doesn't work.

![image](https://github.com/igorski/bitmappery/assets/46873683/6ce5a4e9-6d3c-4f13-a35b-c27305570f9e)

igorski commented:
Well, getting close I guess 😬 

Just to confirm `VITE_S3_GENERATE_THUMBNAILS=true` is defined in your `.env.local`-file right ?
And if you click on any of the image files, they open ?

Ange-Cesari commented:
For both of your question the answer is yes

Do you have any idea of what I could give a try ?

I can do that when I come home

igorski commented:
Hi @Ange-Cesari 

That is most helpful! If you could apply attached .patch file to the latest state of your local repository you should get console logs (info level) for all thumbnail related actions, hopefully this will give us a lead. Given that browsing and opening of files works fine, we should be close 🤞 

[logging-patch.patch](https://github.com/igorski/bitmappery/files/11958081/logging-patch.patch)

Ange-Cesari commented:
> Hi @Ange-Cesari 
> 
> That is most helpful! If you could apply attached .patch file to the latest state of your local repository you should get console logs (info level) for all thumbnail related actions, hopefully this will give us a lead. Given that browsing and opening of files works fine, we should be close 🤞 
> 
> [logging-patch.patch](https://github.com/igorski/bitmappery/files/11958081/logging-patch.patch)
> 

Hey, could you please give the exact guidelines on how to apply the patch (renaming, placing in the exact folder, starting a command...)

So I am sure i'm following your demands

igorski commented:
Hi @Ange-Cesari 

Sure, download attached file (it's plain text) and using CLI navigate to the root folder where you have checked out the BitMappery repository (using the 'aws' branch as before), basically run:

```git apply /path/to/logging-patch.patch`

And you're done. If you run `git diff` you can see that the changes (log entries in `aws-s3-service.ts`) have been applied, but not committed.

After this step you can launch your BitMappery and MinIO containers as normal.

Ange-Cesari commented:
I respected your guidelines and here is a part of the git diff, to show you that it works : 

![image](https://github.com/igorski/bitmappery/assets/46873683/f9a346a6-08f7-4686-85e6-3472844aa89b)
 
![image](https://github.com/igorski/bitmappery/assets/46873683/3e7fc52b-9420-476d-9eb8-15f098ba6228)



![image](https://github.com/igorski/bitmappery/assets/46873683/b8eb0f7a-7bbe-47d2-8390-abb4cc664a8f)
![image](https://github.com/igorski/bitmappery/assets/46873683/e21dfd43-6320-4033-8dea-b00a86cc27ed)
![image](https://github.com/igorski/bitmappery/assets/46873683/6f685e18-d1d8-4f0d-b038-9d3e55e32da6)


![image](https://github.com/igorski/bitmappery/assets/46873683/5296acb3-85a5-4cce-b319-c4b1f5682557)


No matter the changes, no difference unfortunately.

If you have something else to make me try don't hesitate

igorski commented:
Hi @Ange-Cesari

Do you also have screenshots of the DevTools panel with the console tab selected ? I expect to INFO level messages logged there, allowing us to backtrack the issue.

Ange-Cesari commented:
A really strange behaviour actually.

When I open a folder from s3 bucket (by default maybe the last one ?) here it's foo. I then got the error.

Once it's loaded, no problem I can edit the image.

![image](https://github.com/igorski/bitmappery/assets/46873683/9c3e6ae4-6238-4c95-8987-86147c74c685)


At the second I try to import the image from bucket S3, if and only if there is images in the folder, then I get the console error.

I can charge every image, without problem, and the only time it triggers the error is when I import from bucket s3, by selecting a folder where there are images.

And because it's forbidden request, there is no images in the folder, so I need to go back, reselect foo to have access : 

Here is how it looks : 
![image](https://github.com/igorski/bitmappery/assets/46873683/445d5be1-28ab-4e7d-a906-5b4577e3a96c)

igorski commented:
Yeah so (once more) I'm a little stumped. 😬 

I couldn't reproduce this until I started using a Windows 10 device to run the containers, where indeed every other request I got the `RequestTimeTooSkewed` error.

So long story short : 

>S3 uses NTP to keep its system clocks accurate. NTP stands for Network Time Protocol, which provides a standard way of synchronizing computer clocks on servers. NTP can be installed and used to keep the server time accurate, synchronized with accurate atomic clocks located around the world.

BUT there is also the concept of clock drift in Docker containers. I couldn't quite find the details, but apparently suspension state of the host can affect the time in the containers. This is quite strange, especially as MinIO claims they give a lenient time of 15 minutes. I don't see how a drift of 15 mins can happen like that. On my Windows test env, if I got the error, the next request would actually work.

As such, I'm not really sure what is happening here as it seems that this is outside the scope of the BitMappery webapp. I'm not that well versed with setting up Docker containers, maybe this rings a bell to you and you know of a sync setting that can be configured ?

Ange-Cesari commented:
Hey @igorski

As you mentionned, the next request *should* work as expected which is not the case.

Moreover, by building the image you should have run into a new problem which is the version of node used, because you updated something in your code that requires a newer version of the image. I had to go from the 16 to the 18 version, to be able to catch up with your dependencies but that shouldn't be the problem.

Moreover, on macos and windows, the emulation is done with a VM (that's why there is docker desktop only for those one)

I don't think the ntp would cause that, because you should be able to log everything even inside the container and see that there is no desynchronisation. (+ i'm running the app on Windows 11)

I'll try to go deeper into that and see to debug but imo it's the code that it not well interpreted somewhere, like a path, a call back...

I'll try to look at the code and look at the docker issue you mentionned.

If you have any idea in the mean time don't hesitate

igorski commented:
>  I had to go from the 16 to the 18 version, to be able to catch up with your dependencies but that shouldn't be the problem

This could be possible, maybe AWS SDK has stricter requirements. I hadn't noticed as I only run MinIO in the container as it's easier to develop and debug BitMappery's code in my usual dev setup.

Either way, it's puzzling as the end of the day that containerized or not, the clock should be from the same system you're running anything else on.

I will double check the AWS SDK, but request time headers should be set automatically by the SDK, so no idea what's up with that. :/

igorski commented:
Hi @Ange-Cesari 

I could afford some time to look into this again. I managed to replicate the "drift issue" by just having the containers running for a while and then start accessing the S3 storage. I could replicate this with a Win 10 device using MinIO (though BitMappery would not be containerized just for ease of debugging, but apparently the drift also occurs with the host).

Granted, I don't _quite understand_ what the underlying issue with the drift is and most online topics on it are vague or simply state that using ntp should resolve it, but I'm not a Docker hero.

Anyways, I managed to replicate with a local MinIO and could forcefully replicate this with AWS S3 (by simply adjusting my client clock). I've implemented a wrapped execution mechanism that detects the `RequestTimeTooSkewed` issue, and retries the operation (according to S3 spec, the request header time will be honoured). Both MinIO and S3 happily accept the subsequent request.

I'm going to merge this branch into master as during development a ton of improvements with configuration were made, which I'd like to have available to the master branch for further parallel developments.

We can continue discussion (if the issue persists for you) in https://github.com/igorski/bitmappery/issues/25
